cpp/src/replace/replace.cu

```diff
@@ -77,47 +102,97 @@ namespace{ //anonymous
+      const bool new_is_nullable = (new_values.valid != nullptr && new_values.null_count > 0);
+
+      const bit_mask_t* __restrict__ typed_input_valid = reinterpret_cast<bit_mask_t*>(input_col.valid);
+      const bit_mask_t* __restrict__ typed_new_valid = reinterpret_cast<bit_mask_t*>(new_values.valid);
+      bit_mask_t* __restrict__ typed_out_valid = reinterpret_cast<bit_mask_t*>(output.valid);
+
+      gdf_size_type *null_count = nullptr;
+      if (input_is_nullable || new_is_nullable) {
+          RMM_ALLOC(&null_count, sizeof(gdf_size_type), 0);
+          CUDA_TRY(cudaMemset(null_count, 0, sizeof(gdf_size_type)));
+       }
+
+      thrust::device_ptr<const col_type> old_values_begin = thrust::device_pointer_cast(static_cast<const col_type*>(old_values.data));
 
-      const size_t grid_size = nrows / BLOCK_SIZE + (nrows % BLOCK_SIZE != 0);
-      replace_kernel<<<grid_size, BLOCK_SIZE>>>(static_cast<col_type*>(d_col_data),
-                                             nrows,
+      cudf::util::cuda::grid_config_1d grid{input_col.size, BLOCK_SIZE, 1};
+
+      auto replace = replace_kernel<col_type, true, true>;
+
+      if(true == input_is_nullable && false == new_is_nullable){
+        replace = replace_kernel<col_type, true, false>;
+      }
+      else if (false == input_is_nullable && true == new_is_nullable){
+        replace = replace_kernel<col_type, false, true>;
+      }
+      else if (false == input_is_nullable && false == new_is_nullable){
+        replace = replace_kernel<col_type, false, false>;
+      }
+
+      replace<<<grid.num_blocks, BLOCK_SIZE>>>(static_cast<const col_type*>(input_col.data),
+                                             typed_input_valid,
+                                             input_col.size,
+                                             static_cast<col_type*>(output.data),
+                                             typed_out_valid,
+                                             null_count,
                                              old_values_begin,
-                                             old_values_begin + nvalues,
-                                             static_cast<const col_type*>(d_new_values));
+                                             old_values_begin + new_values.size,
+                                             static_cast<const col_type*>(new_values.data),
+                                             typed_new_valid);
+      if (input_is_nullable || new_is_nullable) {
+        CUDA_TRY(cudaMemcpy(&output.null_count, null_count,
+                               sizeof(gdf_size_type), cudaMemcpyDefault));
+        RMM_FREE(null_count, 0);
+      }
     }
   };
 
-  gdf_error find_and_replace_all(gdf_column*       col,
-                                 const gdf_column* old_values,
-                                 const gdf_column* new_values)
+  gdf_column find_and_replace_all(const gdf_column  &input_col,
```

Comment:
Since this is an "internal" API, I'd put it in the `detail` namespace and add a `cudaStream_t` argument that is defaulted to 0.